### PR TITLE
Gtad generating. BaseJob.

### PR DESCRIPTION
**Pull request by Aksem**

### CODE REVIEW

**Review by KitsuneRal — CHANGES_REQUESTED** [review]
Thanks, the direction looks quite sensible; some feedback on the current state below.

File: gtad/typesystem.xml.mustache [hunk 1]
```diff
@@ -6,7 +6,14 @@
...
     {{#operations.operation}}
-    <object-type name="{{camelCaseOperationId}}Job" />
+    <object-type name="{{camelCaseOperationId}}Job">
+      {{#models}}
+      <!-- Inner data structures -->
+        {{#model}}
+      <value-type name="{{name}}" />
+        {{/model}}
+      {{/models}}
+    </object-type>
```
Comment (KitsuneRal) [inline gtad/typesystem.xml.mustache:16]: Haven't tried but that should collapse empty XML elements to make generated files a bit less verbose:
```suggestion
    <object-type name="{{camelCaseOperationId}}Job"{{^models}} />{{/models}}{{#models}}>
      <!-- Inner data structures -->
        {{#model}}
      <value-type name="{{name}}" />
        {{/model}}
    </object-type>
      {{/models}}
```

File: tools/patch_gtag_config.py [hunk 2]
```diff
@@ -17,15 +17,15 @@ def copy_and_patch():
     replace_in_file(
         gtag_config_path,
         original_config_part,
-        '.xml: "{{>../../../gtad/typesystem.xml.mustache}}"\n\n'
+        '.xml: "{{>./typesystem.xml.mustache}}"\n\n'
     )
```
Comment (KitsuneRal) [inline tools/patch_gtag_config.py:21]: You already supply your modified copy of the GTAD config file in the source tree - is this copy-and-patch still needed? If you're concerned about the original file getting updates - I'll make sure to let you know and provide the update to your configuration "fork" as well.

**Review by KitsuneRal — COMMENTED** [review]
Looks much better! A couple of suggestions from me.

File: tools/patch_gtag_config.py (hunk 2 above)
Comment (Aksem) [reply tools/patch_gtag_config.py:21]: Yes, the idea was to be able to update configuration when it's changed in libQuotient. And committed config was only to make sure that we have working version. But if you prefer to keep config and update it manually when needed - no problem, let's do so.

File: gtad/typesystem.xml.mustache (hunk 1 above)
Comment (Aksem) [reply gtad/typesystem.xml.mustache:16]: Yes, it works as expected. Thanks!

File: PyQuotient/event-schemas/schema/m.room.member.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    <load-typesystem name="./../../event-schemas/schema/stripped_state.xml"/>
+    <load-typesystem name="./../../event-schemas/schema/core-event-schema/unsigned_prop.xml"/>
+
+    <namespace-type name="Quotient">
+        <value-type name="SignedData" />
+        <value-type name="Invite" />
+        <value-type name="EventContent" />
+        <value-type name="TheCurrentMembershipStateOfAUserInTheRoom" />
+    </namespace-type>
+</typesystem>
```
[the remark this answers is not in the document]
Comment (Aksem) [reply PyQuotient/event-schemas/schema/m.room.member.xml]: Last version of GTAD configuration doesn't generate this file anymore.

File: tools/patch_gtag_config.py (hunk 2 above)
Comment (KitsuneRal) [reply tools/patch_gtag_config.py]: The thing is - you actually don't have that extensive `gtad.yaml` for your needs, you only use a smaller subset of it (especially when it comes to the Mustache partials library). So yes, let's maintain them separately, at least for now.

**Review by KitsuneRal — COMMENTED** [review]
Ok, I see a lot of progress - great job - but we have to cut this PR somewhere because it starts spinning out of control. It was originally aimed at getting the jobs working, and I totally get that `Connection` might be useful for that; but `Room` should really be spared for another PR. At this point it's worth spending a bit more effort at making the CI working.

File: tests/jobs/test_basejob.py [hunk 4]
```diff
@@ -86,3 +85,137 @@ def test_equal_with_code(self):
...
+    #     base_job = Quotient.BaseJob(
+    #         Quotient.HttpVerb.Get,
+    #         'custom_job',
+    #         '/api/custom/',
+    #         query,
+    #         request_data,
+    #         True
+    #     )
+
+    #     assert isinstance(base_job, Quotient.BaseJob)
```
[the remark this answers is not in the document]
Comment (Aksem) [reply tests/jobs/test_basejob.py:135]: Unfortunately no, even after removing BaseJob::Query it doesn't work. Opened issue https://github.com/quotient-im/PyQuotient/issues/18 .

File: tests/jobs/test_basejob.py [hunk 5]
```diff
@@ -86,3 +85,137 @@ def test_equal_with_code(self):
...
+    def test_json_data(self, base_job):
+        assert isinstance(base_job.json_data, dict)
+
+    def test_json_items(self, base_job):
+        assert isinstance(base_job.json_items, QtCore.QJsonArray)
+    
+    # error doesn't have binding yet. shiboken returns no error related to this property
+    # and doesn't generate a binding
+    # def test_error(self, base_job):
+    #     assert base_job.error == 404
```
[the remark this answers is not in the document]
Comment (Aksem) [reply tests/jobs/test_basejob.py:177]: Opened https://github.com/quotient-im/PyQuotient/issues/19

**Review by KitsuneRal — APPROVED** [review]
Awesome, CI is working - if you wish you can move the `Room` bindings out but since it's not the final PR anyway, it even is ok to keep it.

**Pull request merged**